### PR TITLE
feat(client): support describe source in Java client

### PR DESCRIPTION
**Pull request by vcrfxia**
### Description 

Adds a new method `describeSource(String sourceName)` to the Java client to support `DESCRIBE <STREAM/TABLE>;` statements. Javadocs are included in this PR, usage docs will be added in a follow-up PR.

### Testing done 

Added unit and integration tests.

### Reviewer checklist
- [ ] Ensure docs are updated if necessary. (eg. if a user visible feature is being added or changed).
- [ ] Ensure relevant issues are linked (description should include text like "Fixes #<issue number>")

### CODE REVIEW

File: ksqldb-api-client/src/main/java/io/confluent/ksql/api/client/FieldInfo.java [hunk 1]
```diff
@@ -0,0 +1,38 @@
...
+
+  /**
+   * @return name of this field
+   */
+  String name();
+
+  /**
+   * @return type of this field
+   */
+  ColumnType type();
```
Comment (vcrfxia) [inline ksqldb-api-client/src/main/java/io/confluent/ksql/api/client/FieldInfo.java:31]: The server endpoint actually returns nested field type information in a structured way, so `describeSource` could easily return nested field types, i.e., show the fields and types within arrays, maps, and structs, rather than simply showing "ARRAY", "MAP", and "STRUCT". However, by sharing this `ColumnType` interface that's also used by in the `streamQuery()` and `executeQuery()` methods, we've coupled the two together -- we either have to:
1. Expose nested type information for `describeSource()` and also `streamQuery()`/`executeQuery()`, the latter of which is not as easy to do since the server endpoint used there returns type information as a string rather than in structured format.
1. Do not expose nested type information for either `describeSource()` or `streamQuery()`/`executeQuery()`, as is currently the case in this PR
1. Do not share the `ColumnType` interface between these different client methods.

I don't have a sense of how important it is for users to receive nested type information from `describeSource()`. Unless we think it's important, my vote is option (2) for now.

File: ksqldb-api-client/src/main/java/io/confluent/ksql/api/client/SourceDescription.java [hunk 2]
```diff
@@ -0,0 +1,88 @@
...
+
+/**
+ * Metadata for a ksqlDB stream or table.
+ */
+public interface SourceDescription {
+
+  /**
+   * @return name of this stream/table
+   */
+  String name();
```
Comment (vcrfxia) [inline ksqldb-api-client/src/main/java/io/confluent/ksql/api/client/SourceDescription.java:29]: These method names (`name()`, `type()`, etc.) are a departure from the method names in other Java client interfaces (e.g., https://github.com/confluentinc/ksql/blob/master/ksqldb-api-client/src/main/java/io/confluent/ksql/api/client/QueryInfo.java) in that they aren't prefixed with `get` (`getName()`, `getType()`, etc.). I think it's cleaner to not have the `get` but if we value consistency I can rename these.

File: ksqldb-api-client/src/main/java/io/confluent/ksql/api/client/impl/FieldInfoImpl.java [hunk 3]
```diff
@@ -0,0 +1,76 @@
...
+
+public final class FieldInfoImpl implements FieldInfo {
+
+  private final String name;
+  private final ColumnType type;
+  private final boolean isKey;
+
+  FieldInfoImpl(final String name, final ColumnType type, final boolean isKey) {
+    this.name = Objects.requireNonNull(name);
+    this.type = Objects.requireNonNull(type);
```
Comment (spena) [inline ksqldb-api-client/src/main/java/io/confluent/ksql/api/client/impl/FieldInfoImpl.java:30]: Add the 2nd parameter here. i.e. `requireNonNul(name, "name")`. It helps troubleshoot the NPE errors. I had a problem in the past when I was getting NPE and only a line number appeared. Having the NPE print the String helps easily identify the error instead of a line number.

File: ksqldb-api-client/src/main/java/io/confluent/ksql/api/client/impl/SourceDescriptionImpl.java [hunk 4]
```diff
@@ -0,0 +1,169 @@
...
+    this.type = Objects.requireNonNull(type);
+    this.fields = Objects.requireNonNull(fields);
+    this.topic = Objects.requireNonNull(topic);
+    this.keyFormat = Objects.requireNonNull(keyFormat);
+    this.valueFormat = Objects.requireNonNull(valueFormat);
+    this.readQueries = Objects.requireNonNull(readQueries);
+    this.writeQueries = Objects.requireNonNull(writeQueries);
+    this.timestampColumn = Objects.requireNonNull(timestampColumn);
+    this.windowType = Objects.requireNonNull(windowType);
+    this.sqlStatement = Objects.requireNonNull(sqlStatement);
```
Comment (spena) [inline ksqldb-api-client/src/main/java/io/confluent/ksql/api/client/impl/SourceDescriptionImpl.java:64]: Same comment about using the 2nd parameter to display the variable name on NPE for better troubleshooting. Another reason? say keyFormat is null, it will throw a NPE with line 58. Turns out you saw that exception in ksqlDB 0.12, but master has one new line here, so you'll have to look at 0.12 code to know the right null variable. With the 2nd parameter string will be quickly to find.

**Comment by vcrfxia** [comment]
The new SourceDescription interface introduced in this PR is different from the POJO in ksql-rest-model in subtle, intentional ways, since we don't want to overwhelm client user's with extra information they probably aren't interested in. (The differences for the return types of the other client methods are more pronounced than in this particular case.)

Given that we want to introduce a new interface for the client anyway, are you saying it's still preferable to have the client use Jackson to deserialize the server response into the server's SourceDescriptionEntity first, and then perform a conversion to the client's SourceDescription separately from the JSON parsing? We can definitely do this (assuming your suggestion for shading the Jackson jars works -- I haven't done it before) but I'm failing to see the advantages. If the server's SourceDescriptionEntity evolves, the client won't automatically pick up any newly added fields since they'd have to be manually added into the client's SourceDescription interface in order to be exposed. And the ksqlDB server already has a contract to not make backwards incompatible changes in its REST interfaces, so the manual JSON parsing performed by the client shouldn't break.

**Comment by big-andy-coates** [comment]
> The new SourceDescription interface introduced in this PR is different from the POJO in ksql-rest-model in subtle, intentional ways, 

Can you elaborate?

> since we don't want to overwhelm client users with extra information they probably aren't interested in. (The differences for the return types of the other client methods are more pronounced than in this particular case.)

Why do we assume they're not interested into the info?  If it's not interesting, why would we be returning it from the server?

Can you give examples?

> Given that we want to introduce a new interface for the client anyway

Do you mean the `SourceDescription` interface?  I agree returning interfaces is a good API design and allows us, and users, to easily mock things.  

We can add the interfaces to the models module.

> are you saying it's still preferable to have the client use Jackson to deserialize the server response into the server's SourceDescriptionEntity first, and then perform a conversion to the client's SourceDescription separately from the JSON parsing?

Ideally, they'd be the same thing.  I'm assuming in the fullness of time we want the CLI to use the Java client... right?  We don't want to be managing duplicate code/functionality in the CLI vs Java client.

> We can definitely do this (assuming your suggestion for shading the Jackson jars works -- I haven't done it before) but I'm failing to see the advantages

Advantages are opposite of the disadvantages of using handcrafted JSON parsing:

* much less code
* much less error prone
* shared model with the server
* easier to read and understand.

> If the server's SourceDescriptionEntity evolves, the client won't automatically pick up any newly added fields since they'd have to be manually added into the client's SourceDescription interface in order to be exposed. 

Agreed. Though we should have `SourceDescriptionEntity` implement `SourceDescription` and code reviews should pick up if methods are added to the former and not the latter.

With this approach exposing the new field to the client is simply a case of adding the method to the interface. There is no need to write (and test) verbose JSON parsing code.

> And the ksqlDB server already has a contract to not make backwards incompatible changes in its REST interfaces, so the manual JSON parsing performed by the client shouldn't break.

Famous last words!

**Comment by vcrfxia** [comment]
> Why do we assume they're not interested into the info? If it's not interesting, why would we be returning it from the server?
>
> Can you give examples?

Sure, one example where the client exposes significantly less information is the [CommandStatusEntity](https://github.com/confluentinc/ksql/blob/master/ksqldb-rest-model/src/main/java/io/confluent/ksql/rest/entity/CommandStatusEntity.java), returned from DDL/DML statements. The server's CommandStatusEntity contains loads of information that's inactionable for users of the client:
- commandId can only be used in the context of the `/status` endpoint, which the client doesn't support as it's not particularly useful anyway
- commandStatus ([link](https://github.com/confluentinc/ksql/blob/master/ksqldb-rest-model/src/main/java/io/confluent/ksql/rest/entity/CommandStatus.java#L31)) is not actionable besides knowing whether the command has executed or not, and the query ID which is already extracted by the client
- commandSequenceNumber is no longer applicable for DDL/DML statements since we always wait until the command runner is fully caught up for those, so its only use at the moment is for controlling the flow of admin operations, which is a more niche use case that the client does not currently support
As a result, none of this information is exposed by the client's interface, [ExecutementStatementResult](https://github.com/confluentinc/ksql/blob/master/ksqldb-api-client/src/main/java/io/confluent/ksql/api/client/ExecuteStatementResult.java), at the moment. If we were to expose the entire CommandStatusEntity to client users, it'd be an overwhelming amount of information from which they'd have to figure out how to extract the useful bits, which runs counter to the fact that the client is supposed to be simple to use. Relevant discussion on original KLIP: https://github.com/confluentinc/ksql/pull/5236#discussion_r417935802

> Can you elaborate?

For the particular case of SourceDescriptionEntity, the differences are that the client's interface (SourceDescription):
- does not contain `String statementText` since the point of using the client is that the user doesn't need to know the statement issued was "describe <STREAM/TABLE>;"
- does not contain `List<KsqlWarning> warning` since it's unused (would always be empty)
- un-nests the actual description object from the response, since that's what users would care about
- removes QueryStatusCount from the read/write queries, since those aren't particularly actionable (the client's `listQueries()` method does the same)
- renames `timestamp` to `timestampColumn` to clarify, and makes the type `Optional` to clarify when there is no timestamp column, rather than returning an empty string as the server does
- removes `extended` since it's always false for the client
- removes `statistics` and `errorStats` since those are always empty for the client, because of the point above
- removes `partitions` and `replication` since that seems like a weird thing to describe a source for. Can't think of immediate use cases.
- does not expose nested type info for fields, in order to share the ColumnType interface with what's used by the client for push/pull result rows, though I was on the fence about this one: https://github.com/confluentinc/ksql/pull/5944#discussion_r465789314
- introduces `isKey()` for fields, rather than having an optional `fieldType` which is only ever "KEY" or empty.
- probably some other minor changes I'm forgetting too

As a result of these differences, we can't have SourceDescriptionEntity implement the client's SourceDescription interface (at least without changing the client interface). We have similar problems for the interfaces of other, existing client methods too. For example, the server's `RunningQuery` contains `Set<String> sinks` and `Set<String> sinkKafkaTopics` whereas the client has `Optional<String> sink` and `Optional<String> sinkTopic` to reflect that queries do not (and will never, AFAIA) have multiple sinks.

I think if we could resign the server APIs from scratch, your proposal (have server responses extend client interfaces) makes a lot of sense to me: decide which pieces of information are useful, build an interface out of them, have the server response implement the interface, and have the client send the response directly to the user. Given the current state of our server responses, I think we'd be sacrificing a fair bit of client usability if we were to do this. Maybe it'd be OK if we left out a bunch of the less useful server response fields from the client interfaces? It feels weird to try to shoehorn client interfaces into fitting existing server response formats, though. 

The three things we're trading off between are:
1. maintaining backwards compatibility for the existing server API
1. usability of the client
1. maintenance cost of non-ideal code structure

You've pointed out that the way that the client is currently implemented sacrifices a bit on (3) in favor of (1) and (2). I'm curious to know where you think sacrifices should be made instead. (Or if there's a magic solution I'm not seeing that requires no sacrifices :) )

**Comment by big-andy-coates** [comment]
## Differences:

Let me first just rattle through the differences:

> commandId can only be used in the context of the /status endpoint, which the client doesn't support as it's not particularly useful anyway

Why is it not interesting?  I thought it was needed to know when a command was executed if it hadn't been by the time the response comes back?

Is there any harm in exposing this?  A Java app could always choose to call the /status endpoint itself if it needs to.  Or in the future the client may expose this functionality (or we may remove it).

> commandStatus (link) is not actionable besides knowing whether the command has executed or not, and the query ID which is already extracted by the client

Surely knowing if the command has executed or not is useful info?

> commandSequenceNumber is no longer applicable for DDL/DML statements since we always wait until the command runner is fully caught up for those, so its only use at the moment is for controlling the flow of admin operations, which is a more niche use case that the client does not currently support

Sure, the don't expose it for now. This should be achievable using the interfaces.

> As a result, none of this information is exposed by the client's interface, ExecutementStatementResult, at the moment. If we were to expose the entire CommandStatusEntity to client users, it'd be an overwhelming amount of information from which they'd have to figure out how to extract the useful bits, which runs counter to the fact that the client is supposed to be simple to use.

Then the same argument is true for any one using the rest api directly.  If the model is not good, hard to use, contains useless information, then we should fix the model.

> does not contain String statementText since the point of using the client is that the user doesn't need to know the statement issued was "describe <STREAM/TABLE>;"

Easy enough to support, just don't add the accessor to the interface.

> does not contain List<KsqlWarning> warning since it's unused (would always be empty)

I don't believe that's true.  Well, not if the client supports getting extended information, as that can generate warnings. 

If it's not needed, don't add it to the interfave.

> un-nests the actual description object from the response, since that's what users would care about

Easy enough to support. Add interface method and impl do the unnesting. Marking the method as `@JsonIgnore`.

> removes QueryStatusCount from the read/write queries, since those aren't particularly actionable (the client's listQueries() method does the same)

This were recently added because we want to surface the state of the query.  So not sure why we'd exclude them.  They are actionable, in so much as the user may need to go investigate _why_ a query has failed.

> renames timestamp to timestampColumn to clarify, and makes the type Optional to clarify when there is no timestamp column, rather than returning an empty string as the server does

I'd simple change the model for this one.  We can choose to have the server return `timestamp` and `timestampColumn` for a while to allow people to migrate over if we want.

> removes extended since it's always false for the client

don't add it to the interface then.

> removes statistics and errorStats since those are always empty for the client, because of the point above

don't add it to the interface then.

> removes partitions and replication since that seems like a weird thing to describe a source for. Can't think of immediate use cases.

A source has a topic...  Why not just expose this?    I see no harm in exposing this.

> does not expose nested type info for fields, in order to share the ColumnType interface with what's used by the client for push/pull result rows, though I was on the fence about this one: #5944 (comment)

Yeah, we need to tidy up and standardise how we share schemas across our rest interface for sure.  However, not having column type info in the description is a bit weird!

> introduces isKey() for fields, rather than having an optional fieldType which is only ever "KEY" or empty.

Just change the model to have an `isKey` method!

## Design.

Nothing you've listed IMHO means we should be hand crafting code to handle the JSON when we already have Java classes to do this.  Sure, those classes need some improvements in places, be we can tidy them up as part of exposing the types, (via interfaces), through the client.

We should be planning on using the Java client for the CLI.  Over time making, as the client becomes more fully featured, making it the _only_ way the CLI talks to the server. This is good dogfooding.

In short, we have a model defined in Java, that leverages Jackson serialization. It isn't perfect and could do with improving and streamlining. Doing so improves the Rest interface for all clients.  Writing hand crafted code to parse Json has many downsides. We already have a better way of doing this, so we should use it.  We're just making more work for ourselves now, (as we have to write the code to parse the JSON), and also later when we add to the model, (as we'll have to change not just the model, but also the hand crafted JSON parsing code).

File: ksqldb-api-client/src/main/java/io/confluent/ksql/api/client/impl/FieldInfoImpl.java (hunk 3 above)
Comment (vcrfxia) [reply ksqldb-api-client/src/main/java/io/confluent/ksql/api/client/impl/FieldInfoImpl.java:30]: Done.

File: ksqldb-api-client/src/main/java/io/confluent/ksql/api/client/impl/SourceDescriptionImpl.java (hunk 4 above)
Comment (vcrfxia) [reply ksqldb-api-client/src/main/java/io/confluent/ksql/api/client/impl/SourceDescriptionImpl.java:64]: Done.

**Comment by vcrfxia** [comment]
Discussed with @big-andy-coates offline: We're not going to expose additional information in the client interfaces at this time, in order to stick with the principle of exposing less upfront and exposing information as user demand reveals it's prudent. We are, however, going to refactor the client interfaces into the ksqldb-rest-model module and have the server response classes implement the interfaces, and switch to using Jackson to parse the server responses. We'll shade the Jackson dependency in the client pom to avoid conflicts with other user app dependencies.

The scope of the refactor described above extends far beyond the functionality in this PR so I'm going to merge this PR first (once the build passes) and open a separate PR for the refactor.

**Comment by vcrfxia** [comment]
Tracking follow-up refactor in https://github.com/confluentinc/ksql/issues/6042
